src/main/scala/com/redhat/et/silex/som/som.scala

```diff
@@ -0,0 +1,218 @@
+/*
+ * som.scala
+ * 
+ * author:  William Benton <willb@redhat.com>
+ *
+ * Copyright (c) 2016 Red Hat, Inc.
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.redhat.et.silex.som
+
+import com.redhat.et.silex.util.SampleSink
+
+import breeze.linalg._
+import breeze.numerics._
+
+object Neighborhood {
+  private [this] def gaussian(distance: Double, sigma: Double): Double = math.exp(- (distance * distance) / (sigma * sigma))
+  
+  /** Returns a <tt>dim</tt>-element vector of the values for the one-dimensional Gaussian neighborhood function, centered at <tt>c</tt> */
+  def vec(c: Int, dim: Int, sigma: Double): DenseVector[Double] = {
+    // weights for each distance
+    val weights = (0 to math.max(c, math.abs(dim - c))).map { x => gaussian(x.toDouble, sigma) }
+    DenseVector((0 until dim).map { x => weights(math.abs(x - c))}.toArray)
+  }
+  
+  /** Returns a <tt>xdim</tt>*<tt>ydim</tt>-element matrix of the values for the two-dimensional Gaussian neighborhood function, centered at <tt>xc</tt>, <tt>yc</tt> */
+  def mat(xc: Int, xdim: Int, xsigma: Double, yc: Int, ydim: Int, ysigma: Double): DenseMatrix[Double] = {
+    vec(yc, ydim, ysigma) * vec(xc, xdim, xsigma).t
+  }
+}
+
+class SOM(val xdim: Int, val ydim: Int, val fdim: Int, _entries: DenseVector[DenseVector[Double]], private val mqsink: SampleSink) extends Serializable {
```

Comment:
I was going to suggest generalizing the dimensionality of the map, but it may be more effective to just leave that for some other time and call this SOM2D or something similar, just to capture that it's specific to a 2D SOM